### ISSUE TITLE
Introduce environment specific configurations for Spartacus deployment on ccv2

### ISSUE THREAD

**Issue by tobi-or-not-tobi** [SAP/spartacus#5772]
For angular developers it's very common to use the environments file to distinguish different settings for different builds. This is often used for setting up different tokens for different environment. For example to target a different payment system, analytic system, tag manager, etc. 

The build on Commerce Cloud is currently always using the production build. With this setup, the environment specific settings are not aligned with the build, as production build is always using the production environment settings. 

Different options:
1. We align the angular build with the build to ccv2 environments. This would result in specific builds for specific environments. This will make the builds no longer reusable cross environments. 
2. load environment settings from a file while bootstrapping the app. This would not be recommended for production builds as we introduce a blocking call before we can bootstrap the app (using _config initializers_)
3. bake environment specific variables into UI, similarly to what we've done with the `occ-backend-base-url`. We would use backend properties, which gives the advantage of runtime configuration over build time configuration.

**Comment by neilhubertprice** [comment]
In addition to what Tobias described above, there is also currently a need for a mechanism to differentiate between local development vs deployment in CCv2.  An example currently is the way that the occ-backend-base-url is used.

The options currently are: configure it in code, or use the <meta> injection approach.  If configured in code it overrides the <meta> approach, but the <meta> approach is not available in a local development environment.  It would be possible to achieve this via a standard environment.ts (local dev) vs environment.prod.ts (CCv2) approach ... if there was a standard mechanism for how to handle this in CCv2.  At the moment CCv2 just uses the yarn build, taking the pacakge.json definition (ng build vs ng build --prod).

Most developers will be working in a checked-out version of the same repo used for deployment to CCv2.  The current setup leaves them with only manual file alteration possibilities to make this work.  For a repo that is currently working in CCv2 that means for local development, the developers must either change package.json or environment.ts/environment.prod.ts ... and to enable this, the standard app.module.ts code needs to change slightly (test whether/not the an environment configuration item for occBaseUrl exists).

It seems like some form of option 1 is required for local vs CCv2 builds, as well as perhaps option 3 for applying the CCv2 build across multiple different environments in CCv2

**Comment by aleruz** [comment]
A non trivial implementation may need different configuration properties to set for different environments, for example to connect also to other external systems or for some debugging options. Using environments is at the moment working very well for us.

**Comment by jannesaarinen** [comment]
I would vote for the option 1 from the starters three options.

In the current project we would like to set the server side rendering off for development environment, but true for test and production. Currently it seems that the only option is to create separate builds for development and test / production.

**Comment by alvteren** [comment]
Hi, @tobi-or-not-tobi !
How to use backend properties from Commerce Cloud?

I have disabled the properties on JS Storefront Service.
![image](https://user-images.githubusercontent.com/17045072/92142349-5ae9fb00-ee1c-11ea-9a7c-aa7876b4e596.png)

For other services proreperties aren't disabled

**Comment by alvteren** [comment]
I added properties to JS Storefront Service in Portal Commerce Cloud. 
![image](https://user-images.githubusercontent.com/17045072/92588489-bb0bf180-f2a1-11ea-8d55-52a34581eb40.png)

And I added meta tag with placeholder
![image](https://user-images.githubusercontent.com/17045072/92590626-358a4080-f2a5-11ea-974a-b9bbf394018b.png)

But it doesn't replace placeholder to property value

**Comment by simonkeytree** [comment]
@alvteren - for other services, the properties override back-end configuration properties - the kind of thing you can override on a server-by-server basis in hac.

Have you seen something to indicate that there is a mechanism to include properties into Spartacus?  I don't get to even click the 'SAVE' button if I try this myself, so I just assumed this wasn't available.

**Comment by simonkeytree** [comment]
Hi @Aleruz - how have you got this working in ccv2, if it's always running the production build (I assume using --prod)?

I was kind of expecting that ccv2 backend build process would be calling ng build --configuration=staging for staging, ng build --configuration=production for production, ng build --configuration=dev for development and we could use our own for local development / CI (e.g. ng build --configuration=local) and then it's easy to control config from environment to environment.  Maybe even --configuration=d1 / s1 / p1 as I'm aware some customer might have more environments - e.g. s2 etc)

**Comment by hikmettuysuz** [comment]
Hi @simonkeytree - I have created environment.qa.ts and environment.prod.ts in my project also configurated property in angular.json to replace it. My question is how we can execute our builded code(ng build --configuration=qa ) in cloud?

**Comment by simonkeytree** [comment]
Hi @hikmettuysuz - my question as well. I've raised a support request with SAP to see if there's any kind of workaround. I'm hoping it's not creating three builds that are duplicates of each other other than the config for each environment. Not too bad to do via CI, but a really nasty kludge.

**Comment by neilhubertprice** [comment]
Hi @alvteren, @simonkeytree, @hikmettuysuz 

This is not an official response, but my personal understanding of CCv2 is:
- The basic concept is a single build image per release, and that applies to the Java elements of SAP Commerce as well as the Javascript elements of Spartacus when hosted in CCv2
- For Spartacus this means the same single build is used in all 3 environments, and so it is not picking up the different config files you are talking about.  I think it is probably using the prod file only (but I am not 100% sure)
- The properties that you set in the Commerce Cloud Portal are injected into the server-side processes.  At the moment I don't think there is a mechanism to inject them into the client-side code, which is what you are looking for + what Tobias and I mentioned in earlier comments.

I think all of this is at some level under review by the relevant teams, but I do not know if/how it is progressing.

**Comment by simonkeytree** [comment]
Thanks @neilhubertprice - that's my understanding at the moment. It effectively means you can't use the same build for all three environments, which seems like a big omission as Commerce Cloud officially only supports Spartacus as a SPA, and goes against the approach used for traditional Spring MVC accelerators, where the configuration is well managed through the Manifest file. But the current approach really makes an external CI pipeline essential to ensure git branches have the correct configuration per environment with a build per environment. I can't really think of another good way around it.
There doesn't seem to be any useful environment variables passed from the POD either to indicate whether it's a d1 / s1 / p1 environment, so that's not a possibility either - was a bit surprised by that.

**Comment by neilhubertprice** [comment]
@simonkeytree My view is you should aim to keep the same single build & progress it across environments ... but that means losing out on some of the standard Angular approaches to different environment config.  Elsewhere I have suggested a custom call to the backend SAP Commerce system to retrieve environment-specific configuration - that way you can treat it as data/config separate from the build.

**Comment by simonkeytree** [comment]
Hi @neilhubertprice - I was hoping to workaround it this way. I may be missing something obvious (quite likely!) but how would you know which URL to call to the SAP Commerce system for the specific environment, from the Spartacus pod? I'm looking for something that would identify the environment from within Angular (e.g. an OS environment variable, config from the manifest) but haven't been able to identify anything yet. Any ideas - I'm really hoping I've missed something obvious.

**Comment by neilhubertprice** [comment]
@simonkeytree Each Spartacus instance is served from a particular environment & the OCC url is injected for the API endpoint of that specific environment.  What I suggested above would be an additional (custom) call against that endpoint.  Internal to the Commerce instance it would then have (and respond with) data specific to that environment

**Comment by simonkeytree** [comment]
Hi @nathanss - How is the OCC URL injected? I hadn't realized this got injected from the environment - I thought it was taken from configuration defined in the codebase (either from the meta tag or as defined in a typescript configuration file).

I guess I could take the results from the HOSTNAME environment variable and map this to the specific environment, assuming the hostname remains the same for the environment. That could then be used for using a specific set of configuration values - will give that a try.

Unfortunately, the internal HOSTNAME bears no resemblance to the externally visible hostname, so that was a dead end.

**Comment by neilhubertprice** [comment]
Hi @simonkeytree - I guess that was meant for me rather than Nathan?  Take a look at the 'Caution' box on this page: https://help.sap.com/viewer/b2f400d4c0414461a4bb7e115dccd779/v2005/en-US/63577f67a67347bf9f4765a5385ead33.html or this page in the Spartacus docs: https://sap.github.io/spartacus-docs/configuring-base-url/

**Comment by simonkeytree** [comment]
Hi @neilhubertprice - thanks. Yes apologies to Nathan there. Will give that a go. Bit of a kludge - the standard Angular approach would seem more intuitive, but I guess that's being considered by the Spartacus / Commerce Cloud teams at the moment. Cross fingers.

**Comment by simonkeytree** [comment]
Hi @tobi-or-not-tobi 

What would be the problem with passing (for example)

ng build --configuration=s1 for staging, ng build --configuration=p1 for production, ng build --configuration=d1 for development so that the appropriate config files can be used?

This would seem to eliminate the need to use a meta tag, allow any config you want (per environment), and ensure there's only a single code base.

It does mean duplicating the constants per environment, but that's not the end of the world (and could potentially be worked around by including a common file in each environment .ts file)

Thanks

Simon

**Comment by janwidmer** [comment]
@simonkeytree did you figure something out to make it work via pipeline actions or did you get an answer from the SAP Support request?

**Comment by guiliguili** [comment]
Hi,
Any update on this ?
This is something particularly important to be able to have different {{AuthConfig}} when using 3rd party Identity Provider with different OAuth endpoints and credentials per-environment.

**Comment by janwidmer** [comment]
@guiliguili it looks like, the Cloud is just not supporting any of the ideas described in that ticket. As we also needed this, we came up with the following workaround:

```
export interface IEnvironmentProperties {
	envName: string;
	googleMapsApiKey: string;
}

type TConfigKeys = keyof IEnvironmentProperties;

interface ICloudEnvironment {
	domains: string[];
	envProperties: IEnvironmentProperties;
}

const cloudEnvironments: ICloudEnvironment[] = [
	{
		domains: ['https://api.example.dev.url.com'],
		envProperties: {
			envName: 'DEV',
			googleMapsApiKey: 'my-dev-key',
		},
	},
	{
		domains: ['https://api.example.stage.url.com'],
		envProperties: {
			envName: 'STAGE',
			googleMapsApiKey: 'my-stage-key',
		},
	},
	{
		domains: ['https://api.example.prod.url.com'],
		envProperties: {
			envName: 'PROD',
			googleMapsApiKey: 'my-prod-key',
		},
	},
];

@Injectable({
	providedIn: 'root',
})
export class DynamicConfigurationService {
	dynamicConfiguration: IEnvironmentProperties;

	constructor(private config: OccConfig) {
		const cloudEnvironment =
			cloudEnvironments.find((cloudEnv) => {
				return cloudEnv.domains.includes(this.config.backend?.occ?.baseUrl || '');
			}) || cloudEnvironments[0];

		this.dynamicConfiguration = cloudEnvironment.envProperties;
	}

	public getConfigProperty<TKey extends TConfigKeys>(key: TKey): IEnvironmentProperties[TKey] {
		return this.dynamicConfiguration[key];
	}
}
```

This works, because the only way of detecting the environment in the cloud is currently over the property `this.config.backend?.occ?.baseUrl`.

With this service, it is at least possible to have environment based properties during runtime. 
Downsides are

- Cannot be used outside of runtime
- Config Values are embedded in JS file instead of `environment.ts`

Instead of having the environment properties embedded in the js file, you could also do a custom OCC call to get the properties from somewhere in the backend to prevent having sensitive information embedded in the JS..but as they would then be sent to the client anyway, its almost the same regarding security..

Please note: To make the above Service work, you need to remove all dynamic properties from your `environment.ts` file. Furthermore, you can only have one `index.html` file, where you need to use the meta tag `<meta name="occ-backend-base-url" content="OCC_BACKEND_BASE_URL_VALUE">` where the CCV2 cloud will replace the content attribute with the actual OCC Url depending on the environment. Additionally, you must not have specified the OCC `baseUrl` property in your `storefront.config` as this would again override the value from the html file and not be environment spefic..

**Comment by arminh** [comment]
Hi @tobi-or-not-tobi any updates on this?
We also would like to unify our FE builds for CCv2 and it seems we are not the only ones :)

**Comment by mpern** [comment]
Here is how we should do it:

> ### `index.html` is deployable configuration
> 
>  The HTML document of a single-page application (often `index.html`) is ___not___ static. It varies from environment to environment, and deploy to deploy. The HTML document is a composition of the environment-specific configuration and _immutable_ static assets that define the web application.
> 
> #### `index.html` contains fully-qualified references to the static assets
> 
> ```diff
> - 
> - <script src="main.js" type="text/javascript"></script>
> 
> + 
> + <script src="https://assets.myapp.com/apps/1.0.2/main.js" type="text/javascript"></script>
> ```
> 
> #### `index.html` defines the values of the globally scoped environment variables
> 
> ```html
> <script>
> env = {
>     API: 'https://api.myapp.com',
>     GA: 'UA-126263119-1'
> }
> </script>
> ```
> 

https://immutablewebapps.org/#indexhtml-is-deployable-configuration

**Comment by simonkeytree** [comment]
@mpern - so how would this be implemented? Something like `index-<envname>.html` and if that file exists, that would be the one used for that environment, otherwise fall back to `index.html`?

That would work, and could potentially be done from the Apache config as a rewrite so it's invisible to the application (always looks like index.html), and doesn't need any change in the Spartacus deployment itself.

Actually, strike that, the compiled down version would look very different potentially. Would need a different compiled version for each environment to make this robust unfortunately.

**Comment by neilhubertprice** [comment]
In practice that might not be too far different from the current setup in CCv2.  Currently the build is immutable, but during deploy it parses the index.html to substitute the correct API URL value in place of \<meta name="occ-backend-base-url" content="OCC_BACKEND_BASE_URL_VALUE"\>.  In principal you could do something similar with the script block that Markus was suggesting during the deploy phase.  Actually, this is the same proposition as point 3 in Tobias' original description at the top of this ticket.

**Comment by mpern** [comment]
> Actually, strike that, the compiled down version would look very different potentially. Would need a different compiled version for each environment to make this robust unfortunately.

@simonkeytree - That's where we disagree and that's exactly what the immutable webapp approach warns against (Quote: "immutable static assets", and the production bundle is one of those assets).
The same compiled version is deployed everywhere, `index.html` acts as deployment descriptor and "runtime configuration", in a way. (that's not _that_ different than using multiple `environment.ts`)

We only(?) need:

- first-class, framework-level support for the `env` object by Spartacus
  - generate placeholder `<script>` block with pre-defined `id` and empty `env` object in schemas
  - ~Angular service to read values from `env`~
    EDIT: `environment` should read unknown object properties from `window.env` (or some other fallback mechanism; not sure about SSR though)
  - If you break the `env` object (or `<script>`) block by accident, the build fails
 - CCv2 automation supports `env` (entrypoint script):
   - detect and activate correct `index-<environment>.html` with fallback to `index.html` (great idea!)
   - read/parse `env` object from correct `index.html`, merge object with Service Properties from the CCv2 Cloud Portal and write merged object back to `index.html`
   - highly visible logging of what happend:
     - which `index.html` was used
     - warning if `env` script block/placeholder could not be found in `index.html`
     - before and after values of `env` if it was merged with Service Properties, or info message that no Service Properties required merging
     - the effective/final `env`

Voila!
You have per-environment configuration without an additional request to load the config and you can cache every version of your storefront forever (minus `index.html`, of course.) 

TBH, I would settle for  `index.<environment>.html` support and Service Property injection.

**Comment by simonkeytree** [comment]
I can remember trying out using 'env' when I was originally looking at this, and there wasn't anything useful that would identify the environment (just the pod id and things like that). Just having an environment variable to identify which environment you are in would be nice (and seems kind of basic to me)! Then you could do this for yourself using the methods described earlier in the thread.

Tbh I'm not much of a purist on these things - I just want to have something I can hook into to do what I need.

What we were talking about originally was just being able to identify *which* environment we are in so that this can influence configuration. It would be nice to be able to access the ccv2 front end environment variables as well - as environment variables would be fine.

That being said, it would be nice if there was a structured approach to this in cloud commerce, which is where SAP are guiding everybody to going, with Spartacus being the accelerator of choice these days.

Still, not been on a Spartacus project for a number of months now, so it's all fading faaaasst!

**Comment by janwidmer** [comment]
@neilhubertprice the way how the deployment changes the index.html AFTER the angular build also leads to the problem, that the PWA is not working correctly because the hash of the index.html does not match anymore..

**Comment by neilhubertprice** [comment]
@tobi-or-not-tobi were you aware of this issue around the PWA / hash?  Do we have a solution?

**Comment by CarlEricLavoie** [comment]
@simonkeytree @mpern  - We've faced this issue as well while working on several projects. Overall I agree with @mpern - the application (assets) should be immutable. The way we've fixed it is as follows : 

1. Add a new OCC endpoint to expose site-specific configuration. This allows us to toggle features by environment as well as per site. For instance, GTM configuration keys can differ be site as well as per the environment.
2. Add this new OCC call to the config initializer to ensure the information is readily available.
3. Use it in various services to configure our features.

This change obviously relies on a backend change (OCC) but it feels like the right way to configure our immutable application as a runtime in a "headless storefront" context. It's also not tied to any hosting solution (CCV2, Onprem, etc).

**Comment by balajimohann** [comment]
Hi Team

Any information on when this may be prioritised? It would be great to just use the single build to deploy to multiple environments. Currently

Thanks,
Bala

**Comment by andreas-becker** [comment]
Any updates on this?

**Comment by sam-garland** [comment]
This is something that would be useful in the `@spartacus/epd-visualization` integration library.
We had hoped that customers would be able to use a different value for the `folderUsageId` configuration value for each deployment environment (without creating separate builds for each environment).

**Comment by MarcelLuz** [comment]
This would be more useful if the ccv2 environment supports one of the suggestions in this ticket!
There should be a way to get the environment (d1/s1/p1) especially if you use SSR.

**Comment by sebsauer90** [comment]
It would be awesome to have the possibility to set custom NodeJS variables for build time. Or at least to have a NodeJS variable, which indicates the environment (d1/s1/p1) at build time. Looking forward for any updates on this.

**Comment by andreas-becker** [comment]
On platformServer mode (when page is generated on the server side), you can use `this.windowRef.location.origin` or `window.location.origin` to retrieve the url of the requested call and then determine the specific environment (d1, s1, p1)

**Comment by Pranavbhartia** [comment]
> @simonkeytree @mpern - We've faced this issue as well while working on several projects. Overall I agree with @mpern - the application (assets) should be immutable. The way we've fixed it is as follows :
> 
> 1. Add a new OCC endpoint to expose site-specific configuration. This allows us to toggle features by environment as well as per site. For instance, GTM configuration keys can differ be site as well as per the environment.
> 2. Add this new OCC call to the config initializer to ensure the information is readily available.
> 3. Use it in various services to configure our features.
> 
> This change obviously relies on a backend change (OCC) but it feels like the right way to configure our immutable application as a runtime in a "headless storefront" context. It's also not tied to any hosting solution (CCV2, Onprem, etc).

How did you determine which OCC endpoint needs to be called? d1/s1/p1, etc.. Or are you sending all the config, for all the environments and then checking on UI based on window location href?

**Comment by mpern** [comment]
@Pranavbhartia - the OCC endpoint is the only thing that _does_ get automatically configured per environment.

It works by using the special placeholder `OCC_BACKEND_BASE_URL_VALUE` in your `index.html`.

Details can be found here:
- [Spartacus docs](https://sap.github.io/spartacus-docs/configuring-base-url/#page-title)
- [SAP Commerce Cloud docs @ help.sap.com](https://help.sap.com/viewer/b2f400d4c0414461a4bb7e115dccd779/latest/en-US/63577f67a67347bf9f4765a5385ead33.html)

**Comment by janwidmer** [comment]
@Pranavbhartia you should be able to access the OCC endpoint in your code by using `this.config.backend?.occ?.baseUrl`. With that, you can then do the OCC Backend call to get the environment specific properties as described by @CarlEricLavoie

**Comment by Pranavbhartia** [comment]
@janwidmer and @mpern, thanks for that. I am a bit stuck though. I am a Backend dev trying to resolve this, and not very good with Angular concepts. I have reached to a point where I am able to read the property, but I am just not able to set it to the properties inside the ngModule. Could you please help me out. Please find the code attached.  The conf variable inside the constructor has a value and is printing the full config object that I need. On the other hand, the object before the ngModule has an undefined value and hence it takes the undefined value where I need to provide it with the base url and base site. Instead of going with a backend API, I am using a config file per environment here, so then all the control is within the Spartacus code itself. 

This might not be the best forum for this, but though would help others in my situation as well. 
[spartacus.configuration.module.txt](https://github.com/SAP/spartacus/files/8174607/spartacus.configuration.module.txt)

Edit : Putting the Stackoverflow link for anyone who might be in the same position as me. 
https://stackoverflow.com/questions/71343648/spartacus-access-occ-url-on-the-ngmodule-inside-spartacus-configuration-module

**Comment by janwidmer** [comment]
@Pranavbhartia could you maybe create a stackoverflow question with your code, then the community can help there..

**Comment by timoblume** [comment]
@tobi-or-not-tobi  Is there any hope a solution to this issue will be provided in the future? We are running into the issue of trying to pass environment specific google tag manager Ids in app.module. We still have not found way to get this working. Any help would be great

**Comment by Pranavbhartia** [comment]
Hi @timoblume, 
The only way we could solve it for now was to have separate branches for each environment. We have environment-prod.ts files setup in those branches specific to the environments. I created stackoverflow questions, SAP tickets, spent weeks at trying everything, but nothing really helped. Hopefully there would be a proper solution soon for it, but till then I think that would be the easiest and quickest way to do it. 
Hopefully this helps.

**Comment by sebsauer90** [comment]
Hey, I see there's a lot of frustration about this topic, so I want to share our solution to separate between stages on server side. It's not the nicest way, but the only way we found. We are using a custom storefront, based on express and react, but this also works for spartacus.

So, the only indicator to get information about the stage we found at runtime, is the occ-backend-base-url meta tag in dist/custom-storefront/index.html. If you put `<meta name="occ-backend-base-url" content="OCC_BACKEND_BASE_URL_VALUE" />` in your index.html, the commerce cloud will set it during the build/deployment. So each stage has a separated occ url.

We built a service, which read this meta tag information from the index.html file on server side, to get the current stage. This service is only called once at startup of our nodejs (express) app. After we got the stage, we import asynchronously the corresponding stage configuration.

**Comment by bechte** [comment]
Hey guys,

because SAP doesn't really seem to provide an update on this topic and also no solution for one of the fundamental missing features in Spartacus / Composable Storefront, I thought it would be great to provide an easy-to-integrate solution to the SAP Commerce Cloud community and especially for every project, in which one is force to have different configuration provided, e.g. for single-sign-on with OAuth, etc.

### The solution can be summarized as follows:

- Enhance the Spartacus with a custom configProviderFactory that override the local configuration of the SPA
- Introduction of a new OCC endpoint that provides the frontend configuration in form of a serialized JSON object
- Let the configProviderFactory fetch the configuration from the associate (environment specific) OCC API

As I don't like to do things several times, you can jump-on a open-source extension provided at SAP CX Tools:
https://github.com/sapcxtools/workspace/tree/develop/core-customize/hybris/bin/custom/sapcxtools/sapcxenvconfig

### Steps to integrate

- Download the latest release (4.1.2) at https://github.com/sapcxtools/workspace/releases/tag/v.4.1.2
- Copy over the `sapcxenvconfig` Extension to your project, ideally into your `bin/custom/sapcxtools` folder
- Within the frontend, create the `sapcxenvconfig.ts` file from the template within the `README.md` of the extension
- Add the `securedConfigChunkFromBackend` config factory to your custom module
- Add the `defaultBaseSite` configuration to your `environment.ts` file
- Add your `sapcxenvconfig.frontend.*` properties to your local.properties or environment specific secured configuration as needed

With this solution in place, I think this topic can be marked as solved. Even though it would be great to have the solution provided by SAP... Have fun.

Bechte

**Comment by marcinkapica** [comment]
Hey @bechte,

I have a question to your solution - is it compatible with Spartacus / Composable Storefront running in SSR mode? From what I see in the code it relies on `window.location.protocol` to get the value from `OCC_BASE_URL_META_TAG_NAME` and also requires sending synchronous HTTP request which is not supported in Node.js environment. 

Were you able to make it work in SSR somehow?

**Comment by kaloyan-stankov-cp** [comment]
Hi @marcinkapica , 

Did you find a solution for the multi-environment SSR configuration ? 

I am hitting the same issue currently.

**Comment by bechte** [comment]
yes. we are currently working on a solution to inject the hostname for the api into the initial request to the PWA. with this in place, the resolution with window.location would be obsolete. 

the only thing necessary for this solution would be to have the initial request to the backend node. we solve this by using a reverse proxy that initiates the frontend by a request to the backend on the / request, forwarding the client to a url with parameters. if a client hits a deep link without the configuration set, we redirect to /?targetUrl=… archieving the same. 

still working and testing. but that way we will have support for SSR.

Another solution to less complex setups would be to use the OCC injected URL in the meta information.

**Comment by kaloyan-stankov-cp** [comment]
@bechte or anyone looking at this in the future,

The way we solved issue with multi-environments + SSR is by only using the html tag:
<meta name="occ-backend-base-url" content="OCC_BACKEND_BASE_URL_VALUE" />  but removing the getter in the occConfig - get baseUrl(): string {}.

This way the URL is dynamically injected but without having a direct reference to the window, which causes issues when using SSR.

**Comment by bechte** [comment]
@kaloyan-stankov-cp great idea, I will check, if I can update the example in the README accordingly. Anyway, this will only work in SAP Environments having the external URL configured correctly on them. We sometimes have clients that have multiple domains running the same SAP Commerce Cloud instance behind a proxy. For this scenary, the injection method will not work. But you are right, for 80% of the cases this configuration will work smoothly. Still looking forward to a more general solution. Any ideas are welcome :)
